pennylane/templates/state_preparations/arbitrary_state_preparation.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``ArbitraryStatePreparation`` template.
+"""
+import functools
+import pennylane as qml
+from pennylane.templates.decorator import template
+from pennylane.templates.utils import check_wires, check_shape, get_shape
+
+
+@functools.lru_cache()
+def _state_preparation_pauli_words(num_wires):
+    if num_wires == 1:
+        return ["X", "Y"]
+
+    sub_pauli_words = _state_preparation_pauli_words(num_wires - 1)
+    sub_id = "I" * (num_wires - 1)
+
+    single_qubit_words = ["X" + sub_id, "Y" + sub_id]
+    multi_qubit_words = list(map(lambda word: "I" + word, sub_pauli_words)) + list(
+        map(lambda word: "X" + word, sub_pauli_words)
+    )
+
+    return single_qubit_words + multi_qubit_words
+
+
+@template
+def ArbitraryStatePreparation(angles, wires):
+    """Implements an arbitrary state preparation on the specified wires.
```

Comment:
To me it is kind of clear what an arbitrary unitary is, but I wouldn't know what an arbitrary state prep would be. Maybe explain more? I assume this is ``ArbitraryUnitary`` but with half the parameters?

Reply:
Oh, is it an arbitrary isometry such as from https://arxiv.org/abs/1501.06911?

At first I was a bit confused, wondering if this simply replaces the existing `AmplitudeEmbedding`/`Mottonen`, but from what I can tell, the difference is:

* `AmplitudeEmbedding` directly embeds any given state vector

* `ArbitraryStatePreparation` embeds any given state vector given the some angle parametrization

So the former is good for embedding a known state, while the latter is good for differentiating a arbitrarily parametrized state (where the actual state vector doesn't so much matter)?

Reply:
No, it is just an arbitrary unitary (i.e. an isometry with m=n). 

And you're right with that! I use this if I want to optimize a state over all possible states with the minimum number of parameters. This is of course more of a theorists way of using quantum computers. Usually, you start from layers (i.e. from what you can efficiently do) and build an ansatz class of reachable states out of this. This on the contrary represents a parametrization of any possible state to really cover the whole state space in my optimization.